File: eval/fixtures/nextjs-pms/lib/schemas.ts
import { z } from 'zod'

import { SPACE_KINDS } from '@/lib/space-kinds'

/**
 * Shared validation schemas. Kept free of any database import so they can be
 * used for inline validation in client components AND re-validation inside
 * server actions.
 */

/** Shape of the public booking form, validated on both client and server. */
export const bookingInput = z
  .object({
    guestName: z.string().trim().min(1, 'Please enter your name'),
    email: z.string().trim().email('Enter a valid email'),
    phone: z.string().trim().min(5, 'Enter a valid phone number'),
    checkIn: z.string().min(1, 'Choose a check-in date'),
    checkOut: z.string().min(1, 'Choose a check-out date'),
    partySize: z.coerce.number().int().min(1).max(20),
    notes: z.string().trim().max(1000).optional(),
    /** null = let the front desk assign a space later. */
    spaceId: z.number().int().positive().nullable().default(null),
  })
  .refine((v) => v.checkOut > v.checkIn, {
    message: 'Check-out must be after check-in',
    path: ['checkOut'],
  })

export type BookingInput = z.infer<typeof bookingInput>

export const statusInput = z.object({
  id: z.number().int().positive(),
  status: z.enum(['pending', 'confirmed', 'cancelled']),
})

export const assignInput = z.object({
  id: z.number().int().positive(),
  spaceId: z.number().int().positive().nullable(),
})

export const idInput = z.object({ id: z.number().int().positive() })

/** Shape of the space create / edit form, validated on both client and server. */
export const spaceInput = z.object({
  name: z.string().trim().min(1, 'Give the space a name').max(80),
  kind: z.enum(SPACE_KINDS),
  capacity: z.coerce.number().int().min(1, 'At least 1').max(40),
  /** Nightly rate in whole currency units; blank means "no rate set". */
  rate: z.coerce.number().min(0).max(1_000_000).nullable().default(null),
  notes: z.string().trim().max(500).nullable().default(null),
})

export type SpaceInput = z.infer<typeof spaceInput>

export const spaceUpdateInput = spaceInput.extend({
  id: z.number().int().positive(),
})

export const setSpaceStatusInput = z.object({
  id: z.number().int().positive(),
  status: z.enum(['active', 'archived']),
})

export const availabilityInput = z.object({
  checkIn: z.string().min(1),
  checkOut: z.string().min(1),
  partySize: z.coerce.number().int().min(1),
})

File: eval/fixtures/flask-pms/app/schemas.py
"""Pydantic request models plus the form-coercion helpers they need.

Every form post is parsed through one of these before it reaches the database,
so validation lives in one place and route handlers work with typed data. Flask
hands form values in as strings, so the helpers below turn the raw
``request.form`` values into the ints (and Nones) the models expect.
"""

from typing import Literal

from pydantic import BaseModel, EmailStr, Field, model_validator

from app.space_kinds import SPACE_KINDS

SpaceKind = Literal["room", "suite", "cabin", "villa", "tent", "other"]
ReservationStatus = Literal["pending", "confirmed", "cancelled"]
SpaceStatus = Literal["active", "archived"]


def parse_int(raw: str | None, default: int) -> int:
    """Coerce a form value to int, falling back when it is missing or blank."""
    if raw is None or raw.strip() == "":
        return default
    return int(raw)


def parse_optional_int(raw: str | None) -> int | None:
    """Coerce a form value to int, treating missing / blank as None."""
    if raw is None or raw.strip() == "":
        return None
    return int(raw)


def parse_optional_float(raw: str | None) -> float | None:
    """Coerce a form value to float, treating missing / blank as None."""
    if raw is None or raw.strip() == "":
        return None
    return float(raw)


class BookingInput(BaseModel):
    """The public booking form."""

    guest_name: str = Field(min_length=1)
    email: EmailStr
    phone: str = Field(min_length=5)
    check_in: str = Field(min_length=1)
    check_out: str = Field(min_length=1)
    party_size: int = Field(default=1, ge=1, le=20)
    notes: str | None = Field(default=None, max_length=1000)
    # None = let the front desk assign a space later.
    space_id: int | None = None

    @model_validator(mode="after")
    def check_out_after_check_in(self) -> "BookingInput":
        if self.check_out <= self.check_in:
            raise ValueError("Check-out must be after check-in")
        return self


class StatusInput(BaseModel):
    status: ReservationStatus


class AssignInput(BaseModel):
    space_id: int | None = None


class SpaceInput(BaseModel):
    """The space create / edit form."""

    name: str = Field(min_length=1, max_length=80)
    kind: SpaceKind = "room"
    capacity: int = Field(default=2, ge=1, le=40)
    # Nightly rate in whole currency units; None means "no rate set".
    rate: float | None = Field(default=None, ge=0, le=1_000_000)
    notes: str | None = Field(default=None, max_length=500)

    def as_row(self) -> dict:
        """Column values for a Space, converting the rate to integer cents."""
        return {
            "name": self.name,
            "kind": self.kind if self.kind in SPACE_KINDS else "room",
            "capacity": self.capacity,
            "rate_cents": None if self.rate is None else round(self.rate * 100),
            "notes": self.notes or None,
        }


class SetSpaceStatusInput(BaseModel):
    status: SpaceStatus


class AvailabilityInput(BaseModel):
    check_in: str = Field(min_length=1)
    check_out: str = Field(min_length=1)
    party_size: int = Field(ge=1)
